packages/flutter_tools/lib/src/device.dart

```diff
@@ -294,19 +312,123 @@ abstract class DeviceManager {
 
     return devices;
   }
+}
+
+/// A class for determining how to filter devices based on if they are supported.
+///
+/// If [mustBeSupportedByFlutter] is true, only devices supported by Flutter
+/// will be included.
+///
+/// If [mustBeSupportedForProject] is true, only devices supported by Flutter
+/// and supported for the provided [flutterProject] will be included.
+///
+/// If [mustBeSupportedForAll] is true, only devices supported by Flutter,
+/// supported for the provided [flutterProject], and supported for
+/// `--device all` will be included.
+class DeviceDiscoverySupportFilter {
+  DeviceDiscoverySupportFilter({
+    required this.flutterProject,
+    this.mustBeSupportedByFlutter = false,
+    this.mustBeSupportedForProject = false,
+    this.mustBeSupportedForAll = false,
+  });
```

Comment:
In previous PR, @jmagman asked:
> How about includeDevicesSupportedByFlutter, includeDevicesSupportedByProject, includeUnsupportedDevices or similar.

To which I responded:

> Yeah I had a hard time picking variable names for this. So by default (if you don't set anything) it will include all devices and none will be filtered out. So I thought that using the word include was a bit confusing. For example, if everything is already included, what does includeDevicesSupportedByFlutter really mean? It really means exclude everything that's not supported by flutter. What do you think of excludeDevicesNotSupportedByFlutter, excludeDevicesNotSupportedByProject, excludeDevicesNotSupportedByAll?

Do the `exclude` names make more sense / more inline with our naming conventions?

Reply:
I'll defer to @jmagman on this one, I'm horrible at naming :)